Automattic/wp-calypso
# Jetpack: When taking the checklist tour, pages don't scroll to show the help tooltips

#### Changes proposed in this Pull Request

In the _Plan > My Plan_ page, clicking any item in the _My Checklist_ pane will bring you to the corresponding settings page, but without scrolling to the tour help tooltip. If the tooltip appears below the screen fold line, it leaves users confused about what they should do.

#### Implementation notes

Note that this fix should solve the issue, though the guided tour (or at least the `Step` component) requires more attention, particularly to update its lifecycle.

I first tried to use `lib/scroll-into-view` (a wrapper for `Element.scrollIntoView` and `Element.scrollIntoViewIfNeeded`) as suggested in the code comments. It didn't work out as we'd need to apply it on the card/tooltip directly, which isn't in the flow layout (it's `fixed` positioned).

Here are the 3 parts of the actual fix:
- The init step in `JetpackSignInTour` required the `shouldScrollTo` prop to be passed.
- Even with the prop on, a condition in `UNSAFE_componentWillReceiveProps` prevented the scrolling from occuring. The code is now explicitly checking if the component has scrolled.
- Some properties accessed in `scrollIntoView` in `positioning.ts` weren't defined when the container happened to be the window.

#### Links

Fixes 1164141197617539-as-1165056995892059.

#### Issues

By testing, you might notice a couple of already existing issues:
* Once in the settings page, hitting the back button of the browser will bring you back to _My Plan_, and still show the card/tooltip. #42845
* Depending on the section, a minor scroll jump can happen.
* Not a bug per se: cards don't seem to be shown under a certain viewport width.

#### Testing instructions

* Make sure you have a Jetpack plan
* Make sure you haven't taken the tour yet
* Go to _Plan > My Plan_ page and check the _My Checklist_ section
* Click on the _Try it_ button in the Sign in section
<img width="1026" alt="Screen Shot 2020-05-26 at 9 15 29 AM" src="https://user-images.githubusercontent.com/1620183/82905185-8a41bf80-9f31-11ea-9f90-50e10ebdd4fa.png">

* You're redirected to the corresponding settings page, and shouldn't need to scroll down to see the tooltip (your viewport height shouldn't exceed 800px or so, in order for the scrolling to occur).
<img width="743" alt="Screen Shot 2020-05-26 at 9 17 54 AM" src="https://user-images.githubusercontent.com/1620183/82905376-cf65f180-9f31-11ea-9e5f-b2c177edade1.png">
